File: api/webapi/design/thread.go
package design

import (
	"time"

	. "github.com/shogo82148/goa-v1/design"
	. "github.com/shogo82148/goa-v1/design/apidsl"
)

var _ = Resource("threads", func() {
	Security(JWT, func() {
		Scope("api:access")
	})
	Action("create", func() {
		Routing(POST("threads"))
		Description("スレッドの作成")
		Payload(func() {
			Attribute("text", String, "スレッドの内容", func() {
				Example("テストスレ")
			})
			Attribute("room_id", Integer, "ルームID", func() {
				Example(1)
			})
			Attribute("user_id", Integer, "ユーザーID", func() {
				Example(2)
			})
			Attribute("img", String, "画像", func() {
				Example("https://test.img")
			})
			Required("text", "room_id", "user_id")
		})
		Response(Created, threadUser)
		Response(BadRequest)
		Response(InternalServerError)
	})

	Action("delete", func() {
		Routing(DELETE("threads/:id"))
		Description("スレッドを削除")
		Params(func() {
			Param("id", Integer, "thread id")
		})
		Response(OK)
		Response(NotFound)
		Response(BadRequest)
		Response(InternalServerError)
	})

	Action("get_threads_by_room", func() {
		Routing(GET("threads/room/:id"))
		Description("ルーム内のスレッドを返す")
		Params(func() {
			Param("id", Integer, "Room ID")
			Param("next_id", Integer, "次のID")
		})
		Response(OK, allIndexThreads)
		Response(NotFound)
		Response(InternalServerError)
	})
})

var thread = MediaType("application/vnd.thread", func() {
	Description("スレッド")
	Attribute("id", Integer, "room id")
	Attribute("user_id", Integer, "user id")
	Attribute("room_id", Integer, "room id")
	Attribute("text", String, "スレッド内容")
	Attribute("created_at", DateTime, "作成日", func() {
		Example(time.Date(2019, 01, 31, 0, 0, 0, 0, loc).Format(time.RFC3339))
	})
	Attribute("updated_at", DateTime, "更新日", func() {
		Example(time.Date(2019, 01, 31, 0, 0, 0, 0, loc).Format(time.RFC3339))
	})
	Attribute("img", String, "画像")
	View("default", func() {
		Attribute("id")
		Attribute("user_id")
		Attribute("room_id")
		Attribute("text")
		Attribute("created_at")
		Attribute("updated_at")
		Attribute("img")
	})
	Required("id", "user_id", "room_id", "text", "created_at", "updated_at")
})

var threadUser = MediaType("application/vnd.thread_user", func() {
	Description("スレッドとユーザー情報")
	Attribute("thread", thread, "スレッド")
	Attribute("user", show_user, "ユーザー")
	View("default", func() {
		Attribute("thread")
		Attribute("user")
	})
	Required("thread", "user")
})

var indexThread = MediaType("application/vnd.index_thread", func() {
	Description("スレッドの一覧")
	Attribute("thread_user", threadUser, "スレッドとユーザー")
	Attribute("count_content", Integer, "スレッドの返信数")
	View("default", func() {
		Attribute("thread_user")
		Attribute("count_content")
	})
	Required("thread_user")
})

var allIndexThreads = MediaType("application/vnd.all_index_threads", func() {
	Description("スレッド一覧とNextID")
	Attribute("index_threads", CollectionOf(indexThread), "index_threads")
	Attribute("next_id", Integer, "次取得するThreadID")
	View("default", func() {
		Attribute("index_threads")
		Attribute("next_id")
	})
	Required("index_threads")
})

File: client/lib/repository/thread.ts
import { AllThread, IndexThread, Thread, ThreadUser } from "lib/model/thread";
import { User } from "lib/model/user";
import { asyncApiClient } from "../axios";
import { IndexRoom, Room } from "../model/room";

export const ThreadRepository = {
    getByRoom: async (roomID: number, nextID: number): Promise<AllThread> => {
        const apiClient = await asyncApiClient.create();
        const res = await apiClient.get(`threads/room/${roomID}?next_id=${nextID}`);
        const indexThreads: IndexThread[] = res.data.map((d: any): IndexThread => {
            const thread = d.thread_user.thread;
            const user = d.thread_user.user;
            return {
                threadUser: {
                    thread: new Thread(
                        thread.id,
                        thread.room_id,
                        thread.user_id,
                        thread.text,
                        new Date(thread.created_at),
                        new Date(thread.updated_at),
                    ),
                    user: {
                        id: user.id,
                        name: user.name,
                        createdAt: user.created_at,
                        avatar: user.avatar,
                    },
                },
                countContent: d.count_content,
            };
        });

        return {
            indexThreads: indexThreads,
            nextID: res.data.next_id ? res.data.next_id : null,
        };
    },

    create: async (roomID: number, text: string, img?: string): Promise<ThreadUser> => {
        const apiClient = await asyncApiClient.create();
        const res = await apiClient.post(`threads`, {
            text: text,
            room_id: roomID,
            img: img,
        });
        const thread = res.data.thread;
        const user = res.data.user;

        return {
            thread: {
                id: thread.id,
                roomID: thread.room_id,
                userID: thread.user_id,
                text: thread.text,
                createdAt: new Date(thread.created_at),
                updatedAt: new Date(thread.updated_at),
                img: thread.img,
            },
            user: {
                id: user.id,
                name: user.name,
                createdAt: new Date(user.created_at),
                avatar: user.avatar,
            },
        };
    },
};
